### PR TITLE
Paint Track

### PR DESCRIPTION
**Pull request by duncanspumpkin**
This is an alternative to #2233. The interesting things are in TrackPaint.cpp

### CODE REVIEW

File: src/OpenLoco/src/Paint/PaintTrack.cpp [hunk 1]
```diff
@@ -27,6 +27,389 @@ namespace OpenLoco::Paint
...
+            tpp.boundingBoxSizes[rotation]);
+        session.addToPlotListTrackRoad(
+            baseImage.withIndexOffset(tpp.imageIndexOffsets[rotation][1]),
+            1,
+            heightOffset,
+            tpp.boundingBoxOffsets[rotation] + heightOffset,
+            tpp.boundingBoxSizes[rotation]);
+        session.addToPlotListTrackRoad(
+            baseImage.withIndexOffset(tpp.imageIndexOffsets[rotation][2]),
+            3,
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Paint/PaintTrack.cpp:370]: confirmed that priority 0, 1, 3 is constant across all mergeable track pieces.

File: src/OpenLoco/src/Paint/PaintTrack.cpp [hunk 2]
```diff
@@ -27,6 +27,389 @@ namespace OpenLoco::Paint
...
+            heightOffset,
+            tpp.boundingBoxOffsets[rotation] + heightOffset,
+            tpp.boundingBoxSizes[rotation]);
+
+        session.insertTunnels(height + tpp.tunnelHeights[rotation], trackSession.tunnelType, tpp.tunnelEdges[rotation]);
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Paint/PaintTrack.cpp:375]: mergable track pieces never have a tunnel height but might just keep this anyway.

File: src/OpenLoco/src/Paint/PaintTrack.cpp [hunk 3]
```diff
@@ -27,6 +27,389 @@ namespace OpenLoco::Paint
...
+                rotlSegmentFlags(segment, 3),
+            };
+        }
+        std::array<std::array<uint32_t, 3>, 4> imageIndexOffsets;
+        std::array<World::Pos3, 4> boundingBoxOffsets;
+        std::array<World::Pos3, 4> boundingBoxSizes;
+        std::array<uint8_t, 4> bridgeEdges;
+        std::array<uint8_t, 4> bridgeQuarters;
+        std::array<uint8_t, 4> tunnelEdges;
+        std::array<int16_t, 4> tunnelHeights;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Paint/PaintTrack.cpp:161]: tunnel height is a constant for all rotations.

**Comment by duncanspumpkin** [comment]
So turns out that a lot of what i assumed was wrong. You can't just rotate the edges, segments as tracks that straddle two track id's (such as diagonals) do not have symmetry in that way.

File: src/OpenLoco/src/Paint/PaintTrack.cpp [hunk 4]
```diff
@@ -27,6 +27,389 @@ namespace OpenLoco::Paint
...
+                rotlSegmentFlags(segment, 2),
+                rotlSegmentFlags(segment, 3),
+            };
+        }
+        std::array<std::array<uint32_t, 3>, 4> imageIndexOffsets;
+        std::array<World::Pos3, 4> boundingBoxOffsets;
+        std::array<World::Pos3, 4> boundingBoxSizes;
+        std::array<uint8_t, 4> bridgeEdges;
+        std::array<uint8_t, 4> bridgeQuarters;
+        std::array<uint8_t, 4> tunnelEdges;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Paint/PaintTrack.cpp:160]: Need a bridge type variable (constant for all rotations)

File: src/OpenLoco/src/Paint/PaintTrack.cpp [hunk 5]
```diff
@@ -27,6 +27,389 @@ namespace OpenLoco::Paint
...
+        std::array<int16_t, 4> tunnelHeights;
+        std::array<SegmentFlags, 4> segments;
+        bool isMergable;
+    };
+
+    constexpr std::array<TrackPaintPiece, 1> kStraightTPP = {
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Paint/PaintTrack.cpp:166]: add constants for offsets.

File: src/OpenLoco/src/Paint/PaintTrack.cpp (hunk 4 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/Paint/PaintTrack.cpp]: Lol no its not. Changes with rotation.

File: src/OpenLoco/src/Objects/TrackObject.h [hunk 6]
```diff
@@ -95,47 +95,35 @@ namespace OpenLoco
...
+        constexpr uint32_t kPreviewImage8 = 8;
+        constexpr uint32_t kPreviewImage9 = 9;
+        constexpr uint32_t kPreviewImage10 = 10;
+        constexpr uint32_t kPreviewImage11 = 11;
+        constexpr uint32_t kPreviewImage12 = 12;
+        constexpr uint32_t kPreviewImage13 = 13;
+        constexpr uint32_t kPreviewImage14 = 14;
+        constexpr uint32_t kPreviewImage15 = 15;
+        constexpr uint32_t kPickupFromTrack = 16;
+        constexpr uint32_t kPlaceOnTrack = 17;
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Objects/TrackObject.h:115]: Perhaps clarify that these are UI/button images?

**Comment by AaronVanGeffen** [comment]
Quick testing revealed a few issues happening still:

Screenshot 2024-02-01 at 23 02 57
Screenshot 2024-02-01 at 23 03 05
Screenshot 2024-02-01 at 23 03 10

Save file with viewport set close to the above tracks:

[Randomland by matsku84 - 2050__.SV5.zip](https://github.com/OpenLoco/OpenLoco/files/14131463/Randomland.by.matsku84.-.2050__.SV5.zip)

**Comment by duncanspumpkin** [comment]
> Quick testing revealed a few issues happening still:
> 
> Screenshot 2024-02-01 at 23 02 57 Screenshot 2024-02-01 at 23 03 05 Screenshot 2024-02-01 at 23 03 10
> Save file with viewport set close to the above tracks:
> 
> [Randomland by matsku84 - 2050__.SV5.zip](https://github.com/OpenLoco/OpenLoco/files/14131463/Randomland.by.matsku84.-.2050__.SV5.zip)

argh something went wrong when i rebased this.

**Review by AaronVanGeffen — APPROVED** [review]
Found no more issues in my tests. Very nice work!

**Pull request merged**